google/cloud/storage/_opentelemetry_tracing.py

```diff
@@ -0,0 +1,101 @@
+"""Manages OpenTelemetry tracing span creation and handling."""
+
+import logging
+import os
+
+from contextlib import contextmanager
+
+from google.api_core import exceptions as api_exceptions
+from google.api_core import retry as api_retry
+from google.cloud.storage import __version__
+from google.cloud.storage.retry import ConditionalRetryPolicy
+
+
+ENABLE_OTEL_TRACES_ENV_VAR = "ENABLE_GCS_PYTHON_CLIENT_OTEL_TRACES"
+_DEFAULT_ENABLE_OTEL_TRACES_VALUE = False
+
+enable_otel_traces = os.environ.get(
+    ENABLE_OTEL_TRACES_ENV_VAR, _DEFAULT_ENABLE_OTEL_TRACES_VALUE
+)
+logger = logging.getLogger(__name__)
+
+try:
+    from opentelemetry import trace
+
+    HAS_OPENTELEMETRY = True
+
+    logger.info(f"HAS_OPENTELEMETRY is {HAS_OPENTELEMETRY}")
+except ImportError:
+    logger.debug(
+        "This service is instrumented using OpenTelemetry. "
+        "OpenTelemetry or one of its components could not be imported; "
+        "please add compatible versions of opentelemetry-api and "
+        "opentelemetry-instrumentation packages in order to get Storage "
+        "Tracing data."
+    )
+    HAS_OPENTELEMETRY = False
+
+_default_attributes = {
+    "rpc.service": "CloudStorage",
+    "rpc.system": "http",
+    "user_agent.original": f"gcloud-python/{__version__}",
+}
+
+
+@contextmanager
+def create_span(
+    name, attributes=None, client=None, api_request=None, retry=None, **kwargs
+):
+    """Creates a context manager for a new span and set it as the current span
+    in the configured tracer. If no configuration exists yields None."""
+    if not HAS_OPENTELEMETRY or not enable_otel_traces:
+        yield None
+        return
+
+    tracer = trace.get_tracer(__name__)
+    final_attributes = _get_final_attributes(attributes, client, api_request, retry)
+    # Yield new span.
+    with tracer.start_as_current_span(
+        name=name, kind=trace.SpanKind.CLIENT, attributes=final_attributes
+    ) as span:
+        try:
+            yield span
+        except api_exceptions.GoogleAPICallError as error:
+            span.set_status(trace.Status(trace.StatusCode.ERROR))
+            span.record_exception(error)
+            raise error
```

Comment:
I believe if we "raise" without the "error" it will preserve the traceback without including this bit, which might be more legible.

Reply:
Good call. I’ve updated this, thanks!